code/tfgraph/use.py

```diff
@@ -0,0 +1,56 @@
+# A Tool Generates something from a suite of TensorFlow Graph Recipes
+# (maybe tfgraph-chef)?
+
+import tensorflow as tf
+import importlib
+import argparse
+
+assert tf.__version__ == '1.13.2'
+
+# for each TF graph recipe G in '<DIR>'
+#   if G satisfies '<COND>'
+#   do '<ACTION>'
+#
+# TODO Take "the base directory" as an argument
+# TODO Allow users to select the input graphs
+#   all examples in the directory
+#   selected examples in the directory
+# TODO Allowe
+for example in ['A', 'B']:
+        tf.compat.v1.reset_default_graph()
+        m = importlib.import_module(example)
+
+        print("Generate '" + example + ".pbtxt'")
+        with open(example + ".pbtxt", "w") as f:
+            f.write(str(tf.compat.v1.get_default_graph().as_graph_def(add_shapes=True)))
+        print("Generate '" + example + ".pbtxt' - Done")
+
+        # How to create a TFLITE (F32) from TF Graph
+        #
+        # Reference
+        # - https://www.tensorflow.org/api_docs/python/tf/compat/v1/Session
+        # - https://github.com/tensorflow/docs/blob/r1.13/site/en/api_docs/python/tf/lite/TFLiteConverter.md
+        print("Generate '" + example + ".f32.tflite'")
+        with tf.compat.v1.Session() as sess:
+                converter = tf.lite.TFLiteConverter.from_session(sess, m.inputs, m.outputs)
+                tflite_model = converter.convert()
+                open(example + ".f32.tflite", "wb").write(tflite_model)
+        print("Generate '" + example + ".f32.tflite' - Done")
+
+        # How to create a TensorFlow Lite from T/F Graph
+        #
+        # Reference
+        # - https://github.com/tensorflow/docs/blob/r1.13/site/en/api_docs/python/tf/lite/TFLiteConverter.md
+        # - https://stackoverflow.com/questions/54830869/understanding-tf-contrib-lite-tfliteconverter-quantization-parameters
+        print("Generate '" + example + ".q8.tflite'")
+        with tf.compat.v1.Session() as sess:
+                converter = tf.lite.TFLiteConverter.from_session(sess, m.inputs, m.outputs)
+                # Use 'dummy' quantization
+                converter.inference_type = tf.uint8
+                converter.quantized_input_stats = {}
+                for tensor in m.inputs:
+                        converter.quantized_input_stats[tensor.op.name] = (128, 127)
+                converter.default_ranges_stats = (-128, 127)
+                tflite_model = converter.convert()
+                open(example + ".q8.tflite", "wb").write(tflite_model)
+        print("Generate '" + example + ".q8.tflite' - Done")
```

Comment:
```python
# How to do inference
#
# Reference
# - https://github.com/tensorflow/docs/blob/r1.13/site/en/api_docs/python/tf/lite/Interpreter.md
# - https://www.tensorflow.org/lite/guide/inference
import numpy as np

interpreter = tf.lite.Interpreter(model_path="converted.q8.tflite")
interpreter.allocate_tensors();

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

print(input_details[0])
lhs_shape = input_details[0]['shape']
lhs_data = np.full(lhs_shape, 128, dtype=np.uint8)
lhs_data[0, 1, 1, 0] = 129
print(lhs_data)

interpreter.set_tensor(input_details[0]['index'], lhs_data)

print(input_details[1])
rhs_shape = input_details[1]['shape']
rhs_data = np.full(rhs_shape, 128, dtype=np.uint8)
rhs_data[0, 1, 1, 0] = 129
print(rhs_data)

interpreter.set_tensor(input_details[1]['index'], rhs_data)

interpreter.invoke();

print(output_details[0])
output_data = interpreter.get_tensor(output_details[0]['index'])
print(output_data)
```

Reply:
The example below shows.
- how to write a quantization-aware TF Graph recipe, and
- how to create a quantized TFLITE model from it.

```python
import tensorflow as tf

# How to write a quantization-aware T/F Graph Recipe?
#
# References
# - https://github.com/tensorflow/docs/tree/r1.13/site/en/api_docs/python/tf/quantization
# - https://github.com/tensorflow/docs/blob/r1.13/site/en/api_docs/python/tf/quantization/fake_quant_with_min_max_args.md

lhs_ = tf.placeholder(dtype=tf.float32, shape=(1, 2, 2, 3))
lhs_q_ = tf.fake_quant_with_min_max_args(lhs_, min=0, max=256)
rhs_ = tf.placeholder(dtype=tf.float32, shape=(1, 2, 2, 3))
rhs_q_ = tf.fake_quant_with_min_max_args(rhs_, min=-128, max=127)
out_ = tf.math.add(lhs_q_, rhs_q_)
out_q_ = tf.fake_quant_with_min_max_args(out_, min=-128, max=256)

# Ordered Graph Inputs
inputs = [lhs_q_, rhs_q_]
# Ordered Graph Outputs
outputs = [out_q_]

# How to create a quantized TFLITE model?
#
# Reference
# - https://github.com/tensorflow/docs/blob/r1.13/site/en/api_docs/python/tf/lite/TFLiteConverter.md
# - https://stackoverflow.com/questions/54830869/understanding-tf-contrib-lite-tfliteconverter-quantization-parameters
with tf.compat.v1.Session() as sess:
    converter = tf.lite.TFLiteConverter.from_session(sess, inputs, outputs)
    converter.inference_type = tf.uint8
    
    converter.quantized_input_stats = {}
    for tensor in inputs:
        converter.quantized_input_stats[tensor.op.name] = (128, 1.0) # (zero point, scale)
    
    tflite_model = converter.convert()
    open(tfilte_output_path, "wb").write(tflite_model)
```